### ISSUE TITLE
[ESP-IDF MQTT] repeated disconnection of the device. by "No PING_RESP, disconnected" (IDFGH-10893)

### ISSUE THREAD

**Issue by eric-Kiaofarming** [espressif/esp-idf#12094]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

framework-espidf @ 3.50002.230601

### Operating System used.

macOS

### How did you build your project?

Other (please specify in More Information)

### If you are using Windows, please specify command line type.

None

### Development Kit.

toolchain-xtensa-esp32 @ 11.2.0+2022r1

### Power Supply used.

External 3.3V

### What is the expected behavior?

The device is always online,
And can continue to send and receive data.

### What is the actual behavior?

After several times of sending and receiving Ping actions, there is a possibility that the device may not receive a response and cause the device to be disconnected and reconnected. Reducing the keepalive time can effectively increase the probability of occurrence.

```
        sprintf(mqtt_borker_uri,"mqtt://%s:%d",host,port);

        esp_mqtt_client_config_t mqtt_cfg = {
                .session.protocol_ver = MQTT_PROTOCOL_V_3_1_1,
                .session = {
                        .keepalive = 3,
                        .disable_keepalive = false,
                        .last_will = {
                                .topic = _INIT,
                                .retain = true,
                                .msg = "",
                                .msg_len = 0,
                                .qos = 1,
                        },
                },
        };

        while(ESP_WIFI_CONNECTED == false) {
                 vTaskDelay(50 / portTICK_RATE_MS);
        }

        esp_mqtt_client_handle_t _client = esp_mqtt_client_init(&mqtt_cfg);

        esp_mqtt_client_set_uri(_client, mqtt_borker_uri);
        esp_mqtt_client_register_event(_client, ESP_EVENT_ANY_ID,mqtt_event_handler,NULL);

        esp_mqtt_client_start(_client);
```

### Steps to reproduce.

1. device connection will occur
2. Reducing the interval of keepalive can effectively increase the frequency of occurrence

### Debug Logs.

```plain
D (106239) transport_base: poll_read: select - Timeout before any socket was ready!
D (106239) transport_base: poll_read: select - Timeout before any socket was ready!
D (106239) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
D (106249) mqtt_client: Sent PING successful
D (106259) mqtt_client: mqtt_message_receive: first byte: 0xd0
D (106259) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x0
D (106269) mqtt_client: mqtt_message_receive: total message length: 2 (already read: 2)
D (106279) mqtt_client: mqtt_message_receive: transport_read():2 2
D (106279) mqtt_client: msg_type=13, msg_id=0
D (106289) mqtt_client: MQTT_MSG_TYPE_PINGRESP
D (107299) transport_base: poll_read: select - Timeout before any socket was ready!
D (107299) transport_base: poll_read: select - Timeout before any socket was ready!
D (107299) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
D (108319) transport_base: poll_read: select - Timeout before any socket was ready!
D (108319) transport_base: poll_read: select - Timeout before any socket was ready!
D (108319) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
D (108329) mqtt_client: Sent PING successful
D (109349) transport_base: poll_read: select - Timeout before any socket was ready!
D (109349) transport_base: poll_read: select - Timeout before any socket was ready!
D (109349) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
E (109359) mqtt_client: No PING_RESP, disconnected
D (109369) mqtt_client: Reconnect after 10000 ms
D (109369) event: running post MQTT_EVENTS:2 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (109379) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=2
I (109389) myMQTT: MQTT_EVENT_DISCONNECTED 1
D (119389) mqtt_client: Reconnecting...
D (119389) event: running post MQTT_EVENTS:7 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (119389) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=7
I (119399) myMQTT: MQTT_EVENT_BEFORE_CONNECT
D (119399) esp-tls: host:192.168.31.26: strlen 13
D (119409) esp-tls: [sock=51] Resolved IPv4 address: 192.168.31.26
D (119409) esp-tls: [sock=51] Connecting to server. HOST: 192.168.31.26, Port: 1883
D (119429) mqtt_client: Transport connected to mqtt://192.168.31.26:1883
D (119429) mqtt_client: Sending MQTT CONNECT message, type: 1, id: 0000
D (119499) mqtt_client: mqtt_message_receive: first byte: 0x20
D (119509) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x2
D (119509) mqtt_client: mqtt_message_receive: total message length: 4 (already read: 2)
D (119519) mqtt_client: mqtt_message_receive: read_len=2
D (119519) mqtt_client: mqtt_message_receive: transport_read():4 4
D (119529) mqtt_client: Connected
D (119529) event: running post MQTT_EVENTS:1 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (119539) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=1
I (119549) myMQTT: MQTT_EVENT_CONNECTED
D (119549) mqtt_client: mqtt_enqueue id: 31646, type=8 successful
D (119559) outbox: ENQUEUE msgid=31646, msg_type=8, len=17, size=17
D (119569) mqtt_client: Sent subscribe topic=Y5mJL/PING, id: 31646, type=8 successful
D (119599) mqtt_client: mqtt_message_receive: first byte: 0x90
D (119599) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x3
D (119599) mqtt_client: mqtt_message_receive: total message length: 5 (already read: 2)
D (119609) mqtt_client: mqtt_message_receive: read_len=3
D (119609) mqtt_client: mqtt_message_receive: transport_read():5 5
D (119619) mqtt_client: msg_type=9, msg_id=31646
D (119619) mqtt_client: pending_id=31646, pending_msg_count = 1
D (119629) outbox: DELETED msgid=31646, msg_type=8, remain size=0
D (119639) mqtt_client: deliver_suback, message_length_read=5, message_length=5
D (119639) event: running post MQTT_EVENTS:3 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (119649) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=3
I (119659) myMQTT: MQTT_EVENT_SUBSCRIBED, msg_id=31646
D (119669) mqtt_client: mqtt_enqueue id: 48336, type=3 successful
D (119669) outbox: ENQUEUE msgid=48336, msg_type=3, len=20, size=20
I (119679) MyAgri: mqtt_event_handler => dataRefresh
D (119689) mqtt_client: mqtt_enqueue id: 27541, type=3 successful
D (119689) outbox: ENQUEUE msgid=27541, msg_type=3, len=73, size=93
I (119699) MyAgri: wf8098 00000000,0.000,0,2,0,info,M207,2,1692253651,unsaved[58]
D (119699) mqtt_client: mqtt_message_receive: first byte: 0x40
D (119709) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x2
D (119719) mqtt_client: mqtt_message_receive: total message length: 4 (already read: 2)
D (119729) mqtt_client: mqtt_message_receive: read_len=2
D (119729) mqtt_client: mqtt_message_receive: transport_read():4 4
D (119739) mqtt_client: msg_type=4, msg_id=48336
D (119739) mqtt_client: pending_id=27541, pending_msg_count = 2
D (119749) outbox: DELETED msgid=48336, msg_type=3, remain size=73
D (119759) mqtt_client: received MQTT_MSG_TYPE_PUBACK, finish QoS1 publish
D (119759) event: running post MQTT_EVENTS:5 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (119769) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=5
D (119779) mqtt_client: mqtt_message_receive: first byte: 0x40
D (119789) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x2
D (119789) mqtt_client: mqtt_message_receive: total message length: 4 (already read: 2)
D (119799) mqtt_client: mqtt_message_receive: read_len=2
D (119809) mqtt_client: mqtt_message_receive: transport_read():4 4
D (119809) mqtt_client: msg_type=4, msg_id=27541
D (119819) mqtt_client: pending_id=27541, pending_msg_count = 1
D (119819) outbox: DELETED msgid=27541, msg_type=3, remain size=0
D (119829) mqtt_client: received MQTT_MSG_TYPE_PUBACK, finish QoS1 publish
D (119839) event: running post MQTT_EVENTS:5 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (119849) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=5
D (120869) transport_base: poll_read: select - Timeout before any socket was ready!
D (120869) transport_base: poll_read: select - Timeout before any socket was ready!
D (120869) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
D (121889) transport_base: poll_read: select - Timeout before any socket was ready!
D (121889) transport_base: poll_read: select - Timeout before any socket was ready!
D (121889) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
D (121899) mqtt_client: Sent PING successful
D (121969) mqtt_client: mqtt_enqueue id: 33971, type=3 successful
D (121969) outbox: ENQUEUE msgid=33971, msg_type=3, len=65, size=65
I (121969) MyAgri: wf8098 00000000,0.000,0,2,0,info,M207,2,1692253653[50]
D (121979) mqtt_client: mqtt_message_receive: first byte: 0xd0
D (121979) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x0
D (121989) mqtt_client: mqtt_message_receive: total message length: 2 (already read: 2)
D (121999) mqtt_client: mqtt_message_receive: transport_read():2 2
D (121999) mqtt_client: msg_type=13, msg_id=0
D (122009) mqtt_client: MQTT_MSG_TYPE_PINGRESP
D (122019) mqtt_client: mqtt_message_receive: first byte: 0x40
D (122019) mqtt_client: mqtt_message_receive: read "remaining length" byte: 0x2
D (122029) mqtt_client: mqtt_message_receive: total message length: 4 (already read: 2)
D (122039) mqtt_client: mqtt_message_receive: read_len=2
D (122039) mqtt_client: mqtt_message_receive: transport_read():4 4
D (122049) mqtt_client: msg_type=4, msg_id=33971
D (122049) mqtt_client: pending_id=33971, pending_msg_count = 1
D (122059) outbox: DELETED msgid=33971, msg_type=3, remain size=0
D (122059) mqtt_client: received MQTT_MSG_TYPE_PUBACK, finish QoS1 publish
D (122069) event: running post MQTT_EVENTS:5 with handler 0x400d51d8 and context 0x3ffdcf6c on loop 0x3ffdce50
D (122079) myMQTT: Event dispatched from event loop base=MQTT_EVENTS, event_id=5
D (123099) transport_base: poll_read: select - Timeout before any socket was ready!
D (123099) transport_base: poll_read: select - Timeout before any socket was ready!
D (123099) mqtt_client: esp_mqtt_handle_transport_read_error: transport_read(): call timed out before data was ready!
```

The captured network packets are as follows:

![image](http://www.ilaopo.xyz/test/mqtt.png)

My development environment is as follows:
```
I (29) boot: ESP-IDF 5.0.2 2nd stage bootloader
I (29) boot: compile time 13:56:48
I (29) boot: chip revision: v3.0
I (32) boot.esp32: SPI Speed      : 40MHz
I (37) boot.esp32: SPI Mode       : DIO
I (41) boot.esp32: SPI Flash Size : 8MB
I (46) boot: Enabling RNG early entropy source...
I (516) boot: Loaded app from partition at offset 0x10000
I (516) boot: Disabling RNG early entropy source...
I (528) cpu_start: Pro cpu up.
I (528) cpu_start: Starting app cpu, entry point is 0x40081be0
I (0) cpu_start: App cpu up.
D (536) clk: RTC_SLOW_CLK calibration value: 3321498
I (549) cpu_start: Pro cpu start user code
I (549) cpu_start: cpu freq: 160000000 Hz
I (550) cpu_start: Application information:
I (554) cpu_start: Project name:     controller
I (559) cpu_start: App version:      18b9876-dirty
I (565) cpu_start: Compile time:     Aug 17 2023 13:56:24
I (571) cpu_start: ELF file SHA256:  81f5676df903c66e...
I (577) cpu_start: ESP-IDF:          5.0.2
I (582) cpu_start: Min chip rev:     v0.0
I (586) cpu_start: Max chip rev:     v3.99 
I (591) cpu_start: Chip rev:         v3.0
```
```


### More Information.

My mqtt broker is mosquitto built on the intranet

**Comment by euripedesrocha** [comment]
@eric-Kiaofarming the behavior you are describing is how the keep alive mechanism is supposed to work, refer to [mqtt specification](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349238). 
Reducing the time, you likely increase the disconnection probability, depending on your network situation.

**Comment by eric-Kiaofarming** [comment]
@euripedesrocha  I tested 13 to 30 seconds and had the same problem. I use 3 seconds to speed up the frequency of occurrences.

**Comment by euripedesrocha** [comment]
@eric-Kiaofarming the behavior you described is the expected behavior for MQTT keep alive. You can [disable the keep alive](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/mqtt.html#_CPPv4N24esp_mqtt_client_config_t9session_t17disable_keepaliveE) for the MQTT client If you want to, but the disconnection may happen from the broker side. I suggest that you investigate the adequate value for your system, considering the possible network scenarios. For reference, we set by default the keep alive to 120s, which will make an attempt every 60s.

**Comment by eric-Kiaofarming** [comment]
@euripedesrocha 
Thanks for your reply, I've tried setting 120s and still have this problem.
I have tried disabling keep alive. It works but cannot automatically set offline on mqtt server. I need to create a process to detect it, when the device is offline.

From the network packets I captured, the keepalive problem mostly occurs when "ESP32 inserts other send out (Data or TCP Ack) packets while waiting for ping response", is there any way to prohibit other work intervene while waiting for ping response?
